xbmc/music/MusicDatabase.cpp

```diff
@@ -7082,53 +7082,114 @@ void CMusicDatabase::SetArtForItem(int mediaId, const std::string &mediaType, co
+    if (!BuildSQL(strSQL, filter, strSQL))
+      return false;
+
+    if (!(artistId > 0))
+    {      
+      // Artist ID unknown, so lookup album artist for albums and songs
+      std::string strSQL2;
+      if (albumId > 0)
+      {
+        //Album ID known, so use it to look up album artist(s)
+        strSQL2 = PrepareSQL("SELECT art_id, media_id, media_type, type, "
+          "'albumartist' as prefix, url, album_artist.iOrder as iorder FROM art "
+          "JOIN album_artist ON art.media_id = album_artist.idArtist AND art.media_type ='%s' "
+          "WHERE album_artist.idAlbum = %i ",
+          MediaTypeArtist, albumId);
+        if (bPrimaryArtist)
+          strSQL2 += "AND album_artist.iOrder = 0";
+        else
+          strSQL2 += "ORDER BY iorder";
+
+        strSQL = strSQL + " UNION " + strSQL2;
+      }
+      if (songId > 0)
+      {
+        if (albumId < 0)
+        {
+          //Album ID unknown, so get from song to look up album artist(s)
+          strSQL2 = PrepareSQL("SELECT art_id, media_id, media_type, type, "
+            "'albumartist' as prefix, url, album_artist.iOrder as iorder FROM art "
+            "JOIN album_artist ON art.media_id = album_artist.idArtist AND art.media_type ='%s' "
+            "JOIN song ON song.idAlbum = album_artist.idAlbum  "
+            "WHERE song.idSong = %i ",
+            MediaTypeArtist, songId);
+          if (bPrimaryArtist)
+            strSQL2 += "AND album_artist.iOrder = 0";
+          else
+            strSQL2 += "ORDER BY iorder";
+
+          strSQL = strSQL + " UNION " + strSQL2;
+        }
+
+        // Artist ID unknown, so lookup artist for songs (could be different from album artist)
+        strSQL2 = PrepareSQL("SELECT art_id, media_id, media_type, type, "
+          "'artist' as prefix, url, song_artist.iOrder as iorder FROM art "
+          "JOIN song_artist on art.media_id = song_artist.idArtist AND art.media_type = '%s' "
+          "WHERE song_artist.idsong = %i AND song_artist.idRole = %i ",
+          MediaTypeArtist, songId, ROLE_ARTIST);
+        if (bPrimaryArtist)
+          strSQL2 += "AND song_artist.iOrder = 0";
+        else
+          strSQL2 += "ORDER BY iorder";
+
+        strSQL = strSQL +  " UNION " + strSQL2;
+      }
+    }
+
+    m_pDS2->query(strSQL);
     while (!m_pDS2->eof())
     {
-      art.insert(std::make_pair(m_pDS2->fv(0).get_asString(), m_pDS2->fv(1).get_asString()));
+      ArtForThumbloader artitem;
+      artitem.artType = m_pDS2->fv("type").get_asString();
+      artitem.mediaType = m_pDS2->fv("media_type").get_asString();
+      artitem.prefix = m_pDS2->fv("prefix").get_asString();
+      artitem.url = m_pDS2->fv("url").get_asString();
+      int iOrder = m_pDS2->fv("iorder").get_asInt();
+      // Add order to prefix for multiple artist art for songs and albums e.g. "albumartist2"
+      if (iOrder > 0)
+        artitem.prefix += m_pDS2->fv("iorder").get_asString();
+
+      art.emplace_back(artitem);
```

Comment:
if you are going to do move-semantics, do it all the way, ie. emplace_back({m_pDS-2->fv("type").get_asString(), ... });

Reply:
Unfortunately that would be far less readable, far better with record and named elements IMO